### ISSUE TITLE
npm 5 link / lerna resolve issue

### ISSUE THREAD

**Issue by chrisblossom** [eslint/eslint#9746]
**Tell us about your environment**

* **ESLint Version:** ``4.13.1``
* **Node Version:** ``8.9.3``
* **npm Version:** ``5.6.0``

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Issue:**
When using Lerna with npm v5, eslint is unable to correctly resolve linked packages.

Example Repo: [chrisblossom/eslint-issue-9746](https://github.com/chrisblossom/eslint-issue-9746)

Updated repo to include both scoped and non-scoped example.

**What actually happened? Please include the actual, raw output from ESLint.**
```js
➜  example git:(master) eslint index.js
Cannot find module '@chrisblossom/eslint-config'
Referenced from: /Users/chris/github/eslint-issue-9746/packages/example/.eslintrc.js
Error: Cannot find module '@chrisblossom/eslint-config'
Referenced from: /Users/chris/github/eslint-issue-9746/packages/example/.eslintrc.js
    at ModuleResolver.resolve (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/util/module-resolver.js:74:19)
    at resolve (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config/config-file.js:471:25)
    at load (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config/config-file.js:542:26)
    at configExtends.reduceRight (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config/config-file.js:421:36)
    at Array.reduceRight (<anonymous>)
    at applyExtends (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config/config-file.js:403:28)
    at loadFromDisk (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config/config-file.js:514:22)
    at Object.load (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config/config-file.js:550:20)
    at Config.getLocalConfigHierarchy (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config.js:228:44)
    at Config.getConfigHierarchy (/Users/chris/github/eslint-issue-9746/node_modules/eslint/lib/config.js:180:43)
```

**Comment by platinumazure** [comment]
I'm afraid I don't know much about lerna. Does this issue happen without lerna as well? I'm honestly not sure whether we're supposed to support lerna...

@eslint/eslint-team Can anyone weigh in here?

**Comment by not-an-aardvark** [comment]
This is happening because lerna is installing `eslint` and your scoped config into two different locations. ESLint needs to be able to resolve your shareable config from the location of ESLint itself.

**Comment by davidpelayo** [comment]
@not-an-aardvark how this dependency installation should look like? I'm using lerna and happens to me the same, but if I bypass lerna and install dependencies normally, with `npm`, at the subpackage level, still getting the same error.

**Comment by gugiserman** [comment]
I'm having the same problem and already tried to install locally at package level as well, got the same error output.

**Comment by davidpelayo** [comment]
@gugiserman the solution I came up with a few minutes ago was to avoid hoisting the `eslint` package by installing the dependencies as follow:

```js
lerna bootstrap --hoist --npm-client=npm --nohoist=eslint
```

This resolved correctly the shared eslint config, although it takes more time at installing.

**Comment by davidpelayo** [comment]
I guess this issue can be closed now.

**Comment by ilyavolodin** [comment]
Closing this issue as it seems like it's resolved.

**Comment by chrisblossom** [comment]
I do not think this issue should be closed. I still think this is a bug and eslint should work as expected with npm link/lerna. I've updated [the example repo](https://github.com/chrisblossom/eslint-issue-9746) with a babel example that works as expected.

**Comment by davidpelayo** [comment]
The example @chrisblossom is quite self-explanatory.

As specified [here](https://github.com/lerna/lerna/blob/master/doc/hoist.md#module-resolution), it'd probably has to do with the module resolution, that doesn't follow the [module resolution spec](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders) closely.

**Comment by not-an-aardvark** [comment]
I think this is a general property of libraries that have `peerDependencies`, and isn't specific to ESLint. The current behavior is working as designed.

One might reasonably argue that the design should be changed, and I'd be open to considering that, provided that a clear proposal is presented which takes compatibility issues into account. (Keep in mind that config/plugin resolution has been discussed at length in issues like #3458 -- it might be worth reading some of that discussion if you're interested in learning more about the tradeoffs of ESLint's config resolution.) In any case, this behavior is not a bug.

**Comment by kaicataldo** [comment]
It sounds like we should close this. If someone would like to make a new issue with a concrete proposal, please feel free!

**Comment by SamStonehouse** [comment]
@gugiserman regarding your way of fixing this issue

I have a local package containing my eslint configuration, which is scoped, however when I try and install it in the root directory I get package not found on:

`npm install --save-dev @example/eslint-config-example`

Which isn't surprising as it's not published.

I've also just added it as a dependency in the package.json and attempted `lerna bootstrap` and `lerna bootstrap --hoist` which also don't seem to be able to find the module.

And I've tried simply using `npm link` to add it and that doesn't seem to work either.

In some instances it seems to work without the scoped namespace however I'd like to keep it and you have it in your example, is there any different steps you followed?
